module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php

```diff
@@ -303,6 +316,76 @@ public function getDefaultRequiredDate($checkHolds, $catalog = null,
         return $this->getDateFromArray($dateArray);
     }
 
+    /**
+     * Check if the user-provided dates are valid.
+     *
+     * Returns validated dates and/or an array of validation errors if there are
+     * problems.
+     *
+     * @param string $startDate       User-specified start date
+     * @param string $requiredBy      User-specified required-by date
+     * @param array  $extraHoldFields Hold form fields enabled by
+     * configuration/driver
+     *
+     * @return array
+     */
+    public function validateDates($startDate, $requiredBy, $extraHoldFields)
+    {
+        $result = [
+            'startDateTS' => null,
+            'requiredByTS' => null,
+            'errors' => [],
+        ];
+        if (!in_array('startDate', $extraHoldFields)
+            && !in_array('requiredByDate', $extraHoldFields)
+        ) {
+            return $result;
+        }
+
+        $errors = [];
+        if (in_array('startDate', $extraHoldFields)) {
+            try {
+                $result['startDateTS'] = $startDate
+                    ? $this->dateConverter->convertFromDisplayDate('U', $startDate)
+                    : 0;
+                if ($result['startDateTS'] < time()) {
+                    $errors[] = 'hold_start_date_invalid';
+                }
+            } catch (DateException $e) {
+                $errors[] = 'hold_start_date_invalid';
+            }
+        }
+
+        if (in_array('requiredByDate', $extraHoldFields)) {
+            try {
+                if ($requiredBy) {
+                    $result['requiredByTS'] = $this->dateConverter
+                        ->convertFromDisplayDate('U', $requiredBy, true)
+                        ->setTime(23, 59, 59)
```

Comment:
I guess the need to set a time adds an inconvenient wrinkle to making this more generic, but maybe we could pass timestamp specific flags somehow, either through the input array or a separate options array, to indicate which timestamps need special treatment.